Add support for downloading to seekable files

kyleknap opened the pull request:
This pull request does the following:
* Add a DownloadOutputManager that gives us flexibility in a similar fashion as the UploadInputManager
* Add support for seekable files.

Here is some comparisons:
**Seekable file download**
**Code:**
```py
import time

import boto3
from s3transfer.manager import TransferManager

c = boto3.client('s3')
key = '100MB'

start_time = time.time()
manager = TransferManager(c)
with open(key, 'wb') as f:
    future = manager.download('mybucketfoo', key, f)
    queue_time = time.time()
    print("Time to queue download: %s" % (queue_time - start_time))
    future.result()
    end_time = time.time()
    print("Time to complete download: %s" % (end_time - start_time))
```
**Results:**
```
Time to queue download: 0.00941491127014
Time to complete download: 3.51214289665
```

**Filename**
**Code**
```py
import time

import boto3
from s3transfer.manager import TransferManager

c = boto3.client('s3')
key = '100MB'

start_time = time.time()
manager = TransferManager(c)
future = manager.download('mybucketfoo', key, key)
queue_time = time.time()
print("Time to queue download: %s" % (queue_time - start_time))
future.result()
end_time = time.time()
print("Time to complete download: %s" % (end_time - start_time))
```
**Results**
```
Time to queue download: 0.00508308410645
Time to complete download: 3.74164819717
```

**CLI Comparison**
```
$ time aws s3 cp s3://mybucketfoo/100MB 100MB
download: s3://mybucketfoo/100MB to ./100MB        

real	0m3.738s
user	0m0.844s
sys	0m0.491s
```
cc @jamesls @JordonPhillips

JordonPhillips commented on s3transfer/download.py:
```diff
@@ -38,9 +39,123 @@
+
+    @classmethod
+    def is_compatible(self, download_target):
+        return isinstance(download_target, six.string_types)
```
Maybe you should check here to see if you have write access to the target and validate that it's a valid path.

JordonPhillips commented on tests/unit/test_download.py:
```diff
@@ -65,6 +69,58 @@ def read(self, *args, **kwargs):
+        super(TestDownloadFilenameOutputManager, self).tearDown()
+        shutil.rmtree(self.tempdir)
+
+    def test_is_compatible(self):
```
Could we also test for something that's incompatible, like a `StringIO` or something.

kyleknap replied:
Those would actually be compatible as it is seekable. I will create classes that are not seekable.

JordonPhillips commented on s3transfer/download.py:
```diff
@@ -131,6 +131,27 @@ def _get_temp_fileobj(self):
+    def is_compatible(self, download_target):
+        return (
+            hasattr(download_target, 'seek') and
+            hasattr(download_target, 'tell')
```
An `io.IOBase` will have `seek`, but is not necessarily be seek-able or tell-able. You need to call `io.IOBase.seekable` to be able to tell.

kyleknap replied:
seekable() is not available on the fileobj returned from ``open()`` in Python 2. How about I just check that tell() does not return an IOError?

JordonPhillips replied:
Could you make it a compat function? I feel like we should use it if it's available.

kyleknap replied:
Yeah that fine. I would just need to figure out how ``seekable()`` is implemented in python. I think it is written in C so it may be interesting.

jamesls replied:
Looks like it just tries lseek (seek). https://github.com/python/cpython/blob/master/Modules/_io/fileio.c#L609-L625

kyleknap replied:
Nice. I will do that then.

JordonPhillips commented on tests/functional/test_download.py:
```diff
@@ -145,6 +145,19 @@ def test_download_temporary_file_does_not_exist(self):
         possible_matches = glob.glob('%s*' % self.filename + os.extsep)
         self.assertEqual(possible_matches, [])
 
+    def test_download_for_fileobj(self):
```
Could you add a test for a non-file file-like object, such as `StringIO` or `BytesIO`?

JordonPhillips commented on tests/unit/test_download.py:
```diff
@@ -121,6 +128,40 @@ def test_get_final_io_task(self):
+        self.fileobj.close()
+        super(TestDownloadSeekableOutputManager, self).tearDown()
+
+    def test_is_compatible(self):
```
Could we get some tests for incompatible types, for instance a non-seekable `io.IOBase`

jamesls commented on s3transfer/download.py:
```diff
@@ -38,9 +39,145 @@
+        # writes have finished so done callbacks can be called.
+        return CompleteDownloadNOOPTask(
+            transfer_coordinator=self._transfer_coordinator,
+            is_final=True
```
Is it ever valid to create a `CompleteDownloadNOOPTask` with `is_final=False`.  Should is_final be built into the `CompleteDownloadNOOPTask`.

kyleknap replied:
I can do that.

jamesls commented on s3transfer/download.py:
```diff
@@ -38,9 +39,145 @@
+        raise NotImplementedError('must implement get_fileobj_for_io_writes()')
+
+    def get_final_io_task(self):
+        """Get the final io task to complete the download
```
It would be helpful to document why this is required for subclasses to implement.  The doc string doesn't give a whole lot of information, and I can see myself forgetting why we have to implement this for seekable fileobjs that just return a noop task.

jamesls commented on s3transfer/download.py:
```diff
@@ -38,9 +39,145 @@
     """Task for submitting tasks to execute a download"""
 
+    def _get_download_output_manager_cls(self, transfer_future):
+        """Retieves a class for managing output for a download
```
s/Retieves/Retrieves/

jamesls commented:
Pretty straightfoward, no major issues for me.  I agree with @JordonPhillips's feedback about seekable and test improvements.  Otherwise :shipit:

kyleknap commented:
@jamesls @JordonPhillips 
I updated the following based on feedback:
1) Created a ``seekable()`` function in ``compat.py`` and ported uploads and downloads to use it.
2) Updated docs for ``get_final_io_task`` and made the ``CompleteDownloadNOOPTask`` ``is_final=True`` by default.

Let me know what you think.

jamesls commented on s3transfer/compat.py:
```diff
@@ -50,3 +50,27 @@ def accepts_kwargs(func):
+        # If the fileobj has a seekable attr, try calling the seekable()
+        # method on it.
+        if hasattr(fileobj, 'seekable'):
+            return fileobj.seekable()
```
Can seekable actually raise an IOError/OSError?  Does it need to be under the try?

kyleknap replied:
It does not need to be. I can move it out. It should always return True or False

jamesls commented:
:shipit:, had a question, but otherwise looks good to me.

jamesls commented:
:shipit:

JordonPhillips commented:
```

                       _____|\
                  _.--| SSt |:
                 <____|.----||
                        .---''---,
                         ;..__..'    _...
                       ,'/  ;|/..--''    \
                     ,'_/.-/':            :
                _..-'''/  /  |  \    \   _|/|
               \      /-./_ \;   \    \,;'   \
               ,\    / \:  `:\    \   //    `:`.
             ,'  \  /-._;   | :    : ::    ,.   .
           ,'     ::   /`-._| |    | || ' :  `.`.)
        _,'       |;._:: |  | |    | `|   :    `'
      ,'   `.     /   |`-:_ ; |    |  |  : \
      `--.   )   /|-._:    :          |   \ \
         /  /   :_|   ;`-._;   __..--';    : :
        /  (    ;|;-./_  _/.-:'o |   /     ' |
       /  , \._/_/_./--''/_|:|___|_,'        |
      :  /   `'-'--'----'---------'          |
      | :     O ._O   O_. O ._O   O_.      ; ;
      : `.      //    //    //    //     ,' /
    ~~~`.______//____//____//____//_______,'~
              //    //~   //    //
       ~~   _//   _//   _// ~ _//     ~
     ~     / /   / /   / /   / /  ~      ~~
          ~~~   ~~~   ~~~   ~~~
```
